File: apps/customer_app/lib/services/api/vehicles_service.dart
import 'package:dio/dio.dart';
import '../api/api_client.dart';
import '../../config/api_config.dart';

/// Vehicles service for managing customer vehicles
class VehiclesService {
  final ApiClient _apiClient = ApiClient();

  /// Add vehicle
  Future<Map<String, dynamic>> addVehicle({
    required String type, // 'CAR', 'BIKE', 'TRUCK'
    required String plateNo,
    required String make,
    required String model,
    required int year,
  }) async {
    try {
      final response = await _apiClient.dio.post(
        ApiConfig.vehicles,
        data: {
          'type': type,
          'plateNo': plateNo,
          'make': make,
          'model': model,
          'year': year,
        },
      );
      return response.data;
    } catch (e) {
      throw _handleError(e);
    }
  }

  /// Get all vehicles
  Future<Map<String, dynamic>> getVehicles() async {
    try {
      final response = await _apiClient.dio.get(ApiConfig.vehicles);
      return response.data;
    } catch (e) {
      throw _handleError(e);
    }
  }

  /// Get default vehicle
  Future<Map<String, dynamic>> getDefaultVehicle() async {
    try {
      final response = await _apiClient.dio.get(ApiConfig.vehiclesDefault);
      return response.data;
    } catch (e) {
      throw _handleError(e);
    }
  }

  /// Update vehicle
  Future<Map<String, dynamic>> updateVehicle(
    String vehicleId, {
    String? type,
    String? plateNo,
    String? make,
    String? model,
    int? year,
  }) async {
    try {
      final response = await _apiClient.dio.put(
        ApiConfig.vehicleById(vehicleId),
        data: {
          if (type != null) 'type': type,
          if (plateNo != null) 'plateNo': plateNo,
          if (make != null) 'make': make,
          if (model != null) 'model': model,
          if (year != null) 'year': year,
        },
      );
      return response.data;
    } catch (e) {
      throw _handleError(e);
    }
  }

  /// Delete vehicle
  Future<void> deleteVehicle(String vehicleId) async {
    try {
      await _apiClient.dio.delete(ApiConfig.vehicleById(vehicleId));
    } catch (e) {
      throw _handleError(e);
    }
  }

  String _handleError(dynamic e) {
    if (e is DioException) {
      if (e.response != null) {
        final data = e.response?.data;
        if (data is Map && data.containsKey('message')) {
          return data['message'] as String;
        }
      }
    }
    return e.toString();
  }
}

File: apps/backend/src/vehicles/vehicles.service.ts
import { Injectable, NotFoundException, BadRequestException } from '@nestjs/common';
import { PrismaService } from '../prisma/prisma.service';

@Injectable()
export class VehiclesService {
  constructor(private readonly prisma: PrismaService) {}

  /**
   * Add vehicle for customer
   */
  async addVehicle(userId: string, data: {
    type: string;
    make?: string;
    model?: string;
    year?: number;
    plateNo?: string;
    color?: string;
    isDefault?: boolean;
  }) {
    // If setting as default, unset other defaults
    if (data.isDefault) {
      await this.prisma.customerVehicle.updateMany({
        where: { userId, isDefault: true },
        data: { isDefault: false },
      });
    }

    return this.prisma.customerVehicle.create({
      data: {
        userId,
        ...data,
      },
    });
  }

  /**
   * Update vehicle
   */
  async updateVehicle(vehicleId: string, userId: string, data: {
    type?: string;
    make?: string;
    model?: string;
    year?: number;
    plateNo?: string;
    color?: string;
    isDefault?: boolean;
  }) {
    const vehicle = await this.prisma.customerVehicle.findUnique({
      where: { id: vehicleId },
    });

    if (!vehicle) throw new NotFoundException('Vehicle not found');
    if (vehicle.userId !== userId) {
      throw new BadRequestException('Access denied');
    }

    // If setting as default, unset other defaults
    if (data.isDefault) {
      await this.prisma.customerVehicle.updateMany({
        where: { userId, isDefault: true, id: { not: vehicleId } },
        data: { isDefault: false },
      });
    }

    return this.prisma.customerVehicle.update({
      where: { id: vehicleId },
      data,
    });
  }

  /**
   * Get customer vehicles
   */
  async getCustomerVehicles(userId: string) {
    return this.prisma.customerVehicle.findMany({
      where: { userId },
      orderBy: [{ isDefault: 'desc' }, { createdAt: 'desc' }],
    });
  }

  /**
   * Get default vehicle
   */
  async getDefaultVehicle(userId: string) {
    return this.prisma.customerVehicle.findFirst({
      where: { userId, isDefault: true },
    });
  }

  /**
   * Delete vehicle
   */
  async deleteVehicle(vehicleId: string, userId: string) {
    const vehicle = await this.prisma.customerVehicle.findUnique({
      where: { id: vehicleId },
    });

    if (!vehicle) throw new NotFoundException('Vehicle not found');
    if (vehicle.userId !== userId) {
      throw new BadRequestException('Access denied');
    }

    await this.prisma.customerVehicle.delete({
      where: { id: vehicleId },
    });

    return { ok: true };
  }
}

